### PR TITLE
Refactor the command parser

### PR DESCRIPTION
**Pull request by Half-Shot**
We wrote a command parser twice, and it's about time we reconciled it a bit.

### CODE REVIEW

File: src/bridge/AdminRoomHandler.ts [hunk 1]
```diff
@@ -127,7 +149,7 @@ const COMMANDS: {[command: string]: Command|Heading} = {
         requiresPermission: CommandPermission.Admin,
     },
     '!unlink': {
-        example: "!unlink !room:example.com irc.example.net #foobar",
+        example: "unlink !room:example.com irc.example.net #foobar",
```
Comment (progval) [inline src/bridge/AdminRoomHandler.ts]: looks like you forgot to remove the `!` prefix from `!plumb` and `!unlink`, and removed it from the example instead

Comment (Half-Shot) [reply src/bridge/AdminRoomHandler.ts]: ta, this'll be because we decided to use different forms of quote.

File: src/bridge/AdminRoomHandler.ts [hunk 2]
```diff
@@ -35,6 +36,27 @@ enum CommandPermission {
...
+    if (typeof event.content?.body !== "string") {
+        return null;
+    }
+    // First line only
+    const body = event.content.body.trim().split("\n")[0];
+    if (!body.startsWith(prefix)) {
+        return null;
+    }
+    // Assumes all commands have the form "!wxyz [irc.server] [args...]"
+    const [cmd, ...args] = body.slice(prefix.length).split(" ");
```
Comment (AndrewFerr) [inline src/bridge/AdminRoomHandler.ts:56]: Should this check if `cmd` ends up being empty?  i.e. should `null` be returned for an `event.content.body` of `"!"`?

File: src/provisioning/Provisioner.ts [hunk 3]
```diff
@@ -860,9 +870,15 @@ export class Provisioner extends ProvisioningApi {
 
         const ircDomain = body.remote_room_server;
         const ircChannel = body.remote_room_channel;
+
+
```
Comment (AndrewFerr) [inline src/provisioning/Provisioner.ts]: Nitpick, is one newline sufficient? :slightly_smiling_face:

File: src/bridge/AdminRoomHandler.ts (hunk 2 above)
Comment (Half-Shot) [reply src/bridge/AdminRoomHandler.ts:56]: Yes! Good point.

File: src/provisioning/Provisioner.ts (hunk 3 above)
Comment (Half-Shot) [reply src/provisioning/Provisioner.ts]: It looks like I do more work! But yep, removed.

**Review by AndrewFerr — APPROVED** [review]

File: src/models/IrcRoom.ts [hunk 4]
```diff
@@ -18,6 +18,17 @@ import { RemoteRoom } from "matrix-appservice-bridge";
...
 
+/**
+ * Validate that a channel name is in the correct form.
+ * NOTE: This accepts both # and & as channel prefixes.
+ * @param channel Channel name
+ * @returns True if the channel is valid, or false otherwise.
+ */
+export function validateChannelName(channel: string) {
+    // According to https://www.rfc-editor.org/rfc/rfc1459#section-1.3
+    return /^(#|&)([^\s\cG,]){1,199}$/.test(channel);
```
Comment (progval) [inline src/models/IrcRoom.ts]: I don't think AS-irc supports `&`-channels, so you can probably disallow those.

(also non-space whitespaces are technically allowed in practice, eg. on OFTC, but that's just asking for trouble so it's fine to disallow them too)

Comment (Half-Shot) [reply src/models/IrcRoom.ts]: I left them in on a RFC technicality, but tbh it's probably not useful. :+1: 

and uuuurgghhhh whitespaces are something we'll care about later if people are desperate.

Comment (Half-Shot) [reply src/models/IrcRoom.ts]: `\cG matches the control sequence CTRL+G (ASCII 7)` according to my notes.

**Pull request merged**